File: identhub/src/main/java/de/solarisbank/identhub/file/FileController.java
package de.solarisbank.identhub.file;

import android.content.Context;

import java.io.File;
import java.io.IOException;

import de.solarisbank.identhub.base.Optional;
import okio.BufferedSink;
import okio.BufferedSource;
import okio.Okio;

public class FileController {
    private final String pdfDirectoryPath;

    public FileController(Context context) {
        pdfDirectoryPath = context.getFilesDir() + "/pdf";
    }

    public Optional<File> save(String fileName, BufferedSource bufferedSource) throws IOException {
        File pdfDirectory = new File(pdfDirectoryPath);
        pdfDirectory.mkdirs();

        File destinationFile = new File(pdfDirectoryPath, fileName);
        try (BufferedSink sink = Okio.buffer(Okio.sink(destinationFile))) {
            sink.writeAll(bufferedSource);
            return Optional.of(destinationFile);
        } catch (IOException io) {
            return Optional.empty();
        }
    }

    public Optional<File> get(String fileName) {
        File destinationFile = new File(pdfDirectoryPath, fileName);
        if (destinationFile.exists()) {
            return Optional.of(destinationFile);
        }
        return Optional.empty();
    }
}

File: identhub/src/main/java/de/solarisbank/identhub/file/FileController.kt
package de.solarisbank.identhub.file

import android.content.Context
import de.solarisbank.identhub.file.OkioKotlinAdapter.fillFileIn
import okio.BufferedSource
import java.io.File
import java.io.IOException

class FileController(context: Context) {
    private val pdfDirectoryPath: String = context.filesDir.toString() + "/pdf"

    @Throws(IOException::class)
    fun save(fileName: String, bufferedSource: BufferedSource): File? {
        val pdfDirectory = File(pdfDirectoryPath)
        pdfDirectory.mkdirs()
        val destinationFile = File(pdfDirectoryPath, fileName)
        return fillFileIn(bufferedSource, destinationFile)
    }

    fun retrieve(fileName: String): File? {
        val destinationFile = File(pdfDirectoryPath, fileName)
        return if (destinationFile.exists()) {
            destinationFile
        } else null
    }

}